compositor_render/src/transformations/web_renderer/browser.rs

```diff
@@ -0,0 +1,73 @@
+use compositor_chromium::cef;
+use compositor_common::scene::Resolution;
+use crossbeam_channel::{Receiver, Sender};
+
+pub(super) struct BrowserState {
+    frame_rx: Receiver<Vec<u8>>,
+    frame_data: Vec<u8>,
+}
+
+impl BrowserState {
+    pub fn new(frame_rx: Receiver<Vec<u8>>) -> Self {
+        Self {
+            frame_rx,
+            frame_data: Vec::new(),
+        }
+    }
+
+    pub fn retrieve_frame(&mut self) -> &[u8] {
+        if let Ok(frame) = self.frame_rx.try_recv() {
+            self.frame_data = frame;
+        }
+
+        &self.frame_data
+    }
+}
+
+pub struct BrowserClient {
+    frame_tx: Sender<Vec<u8>>,
+    resolution: Resolution,
+}
+
+impl cef::Client for BrowserClient {
+    type RenderHandlerType = RenderHandler;
+
+    fn render_handler(&self) -> Option<Self::RenderHandlerType> {
+        Some(RenderHandler::new(self.frame_tx.clone(), self.resolution))
+    }
+}
+
+impl BrowserClient {
+    pub fn new(frame_tx: Sender<Vec<u8>>, resolution: Resolution) -> Self {
+        Self {
+            frame_tx,
+            resolution,
+        }
+    }
+}
+
+pub struct RenderHandler {
+    frame_tx: Sender<Vec<u8>>,
+    resolution: Resolution,
+}
+
+impl cef::RenderHandler for RenderHandler {
+    fn resolution(&self, _browser: &cef::Browser) -> Resolution {
+        self.resolution
+    }
+
+    fn on_paint(&self, _browser: &cef::Browser, buffer: &[u8], _resolution: Resolution) {
+        if !self.frame_tx.is_full() {
```

Comment:
- This might not be bug if chrome has some guarantees about how on_paint is called, but regardless of that we should not be relying on logic like that. The problem I see here is that is_full can return false, but in the meantime other thread could write sth to that channel and this write would block.
- The other issue is that you are not sending the latest possible frame. This would probably give at most 1 frame, but it's not necessary if chrome is able to keep up, so we should not do it.

You could create an unbounded channel and when reading frames in the render drain it entirely and use last frame